### PR TITLE
Update DG for undo and redo commands

### LINKED ISSUE #112: Add sequence diagram for undo command to DG
Add a sequence diagram to model the `undo` command.

### PR DESCRIPTION
**Pull request by bryanyee33**
Closes #112

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -209,40 +213,97 @@ Similarly, how an undo operation goes through the `Model` component is shown bel
...
+      * Pointers have to be carefully implemented.
+</li>
+
+* **Alternative 3:** LinkedList.
+    * Pros:
+      * Fast in dropping many nodes after a specified index.
+      * Simple data structure.
+    * Cons:
+      * Time-consuming to implement: Unfortunately, the built-in LinkedList does not have a method to drop all nodes after a certain index,
+        and hence a custom data structure would have to be used.
```
Comment (aureliony) [inline docs/DeveloperGuide.md:305]: Why would a custom data structure be necessary? The developer can call `linkedList.removeLast()` until the required size.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -209,40 +213,97 @@ Similarly, how an undo operation goes through the `Model` component is shown bel
...
+
+* **Alternative 2:** Save the entire address book into storage as JSON files.
+    * Pros:
+      * Saves history between different program launches.
+    * Cons:
+      * May have performance issues with many storage accesses.
+      * Increases coupling as `Model` will now need a reference to `Storage`.
+</li>
+
+* **Alternative 3:** Save the entire address book as a JSON `String`.
```
Comment (aureliony) [inline docs/DeveloperGuide.md:262]: Should this mention "as a JSON in RAM"?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -209,40 +213,97 @@ Similarly, how an undo operation goes through the `Model` component is shown bel
...
+    * Pros:
+      * Reduces memory footprint as compared to objects.
+      * Faster than JSON files as there are no accesses to storage.
+    * Cons:
+      * May have performance issues as it has to be deserialized each time.
+</li>
+
+* **Alternative 4:** Individual command knows how to undo/redo by itself.
```
Comment (aureliony) [inline docs/DeveloperGuide.md:270]: Should this be changed to "Each command implements the `undo()` and `redo()` methods"?

**Review by aureliony — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (bryanyee33) [reply docs/DeveloperGuide.md:305]: The reasoning to use a linked list instead was to possibly make removing all nodes after a certain index O(1).
Using linkedList.removeLast() multiple times would make it O(N), resulting in no benefits of using a linked list as compared to other data structures.
I will update the description in the DG

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (bryanyee33) [reply docs/DeveloperGuide.md:262]: Yes, that would be clearer. I will update it.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (bryanyee33) [reply docs/DeveloperGuide.md:270]: I feel like that would be less clear in terms of phrasing.
I'll switch it to "Each command implements their own specific `undo()` and `redo()` methods"

**Review by aureliony — APPROVED** [review]

**Pull request merged**